### PR TITLE
fix: Secure storage memory leak

### PR DESCRIPTION
**Pull request by BiancaMihaila**
- Amend sharedPrefStore to accept Context(Application Context)
- Amend Secure Store methods to take in context (FragmentActivity/ Activity Context) to allow yo be injected via Hilt
- Amend Authenticator to get the context and close it once method has been invoked and once completed

### CODE REVIEW

File: app/src/main/java/uk/gov/android/securestore/SharedPrefsStore.kt [hunk 1]
```diff
@@ -23,31 +23,37 @@ class SharedPrefsStore(
...
         }
     }
 
-    override fun delete(key: String) {
+    override fun delete(key: String, context: FragmentActivity) {
         writeToPrefs(key, null)
+        authenticator.init(context)
         try {
             cryptoManager.deleteKey()
+            authenticator.close(context)
```
Comment (chrisdondig) [inline app/src/main/java/uk/gov/android/securestore/SharedPrefsStore.kt:45]: could you move this line to a finally block please?

File: app/src/main/java/uk/gov/android/securestore/SharedPrefsStore.kt [hunk 2]
```diff
@@ -57,6 +63,7 @@ class SharedPrefsStore(
                         authPromptConfig
                     )
                 } ?: continuation.resumeWith(Result.success(null))
+                authenticator.close(context)
```
Comment (chrisdondig) [inline app/src/main/java/uk/gov/android/securestore/SharedPrefsStore.kt:66]: could you also move this line to a finally block please?

File: app/src/main/java/uk/gov/android/securestore/authentication/UserAuthenticator.kt [hunk 3]
```diff
@@ -38,4 +42,8 @@ internal class UserAuthenticator(
             AccessControlLevel.PASSCODE_AND_CURRENT_BIOMETRICS ->
                 BIOMETRIC_STRONG or DEVICE_CREDENTIAL
         }
+
+    override fun close(context: Context) {
+        fragmentContext.finish()
```
Comment (chrisdondig) [inline app/src/main/java/uk/gov/android/securestore/authentication/UserAuthenticator.kt:47]: As discussed, make fragmentContext nullable at declaration and set it to null here

Comment (BiancaMihaila) [reply app/src/main/java/uk/gov/android/securestore/authentication/UserAuthenticator.kt]: Done

File: app/src/main/java/uk/gov/android/securestore/SharedPrefsStore.kt (hunk 2 above)
Comment (BiancaMihaila) [reply app/src/main/java/uk/gov/android/securestore/SharedPrefsStore.kt]: Done, thank you

File: app/src/main/java/uk/gov/android/securestore/SharedPrefsStore.kt (hunk 1 above)
Comment (BiancaMihaila) [reply app/src/main/java/uk/gov/android/securestore/SharedPrefsStore.kt]: Done, thank you

**Review by chrisdondig — DISMISSED** [review]

File: app/src/main/java/uk/gov/android/securestore/authentication/Authenticator.kt [hunk 4]
```diff
@@ -13,9 +14,14 @@ interface Authenticator {
      * @param configuration Configuration for the UI elements of the OS Biometric Prompt
      * @param handler An [AuthenticatorCallbackHandler] to handle success, failure and error states from authentication
      */
+
```
Comment (flavius-mester) [inline app/src/main/java/uk/gov/android/securestore/authentication/Authenticator.kt:17]: I would move this up as not to confuse the method docs

File: app/src/main/java/uk/gov/android/securestore/SharedPrefsStore.kt [hunk 5]
```diff
@@ -23,31 +23,39 @@ class SharedPrefsStore(
 ) : SecureStore {
     private val sharedPrefs = context.getSharedPreferences(configuration.id, Context.MODE_PRIVATE)
 
-    override suspend fun upsert(key: String, value: String): String {
+    override suspend fun upsert(key: String, value: String, context: FragmentActivity): String {
+        authenticator.init(context)
```
Comment (flavius-mester) [inline app/src/main/java/uk/gov/android/securestore/SharedPrefsStore.kt:27]: I think it's a good proactice to move the `init` into the `try`.

You `init` it in the `try`, and you close it in `finally`

**Review by flavius-mester — APPROVED** [review]

**Pull request merged**